### PR TITLE
Add option `throw_error_with_invalid_path` to Project Setting and remove cache error from `AnimationTree`

### LINKED ISSUE #65608: AnimationTree infinite loop, ERROR: Condition "!track_cache.has(path)" is true. Continuing.
### Godot version

4.0-alpha16

### System information

PopOS 22.04, 5.19.0-76051900-generic kernel, Ryzen 6800h, Radeon 6700m, amdgpu, Vulkan

### Issue description

The whole application started to lag and I noticed this output over and over in the terminal:
```
ERROR: Condition "!track_cache.has(path)" is true. Continuing.
   at: _process_graph (scene/animation/animation_tree.cpp:990)
ERROR: Condition "!track_cache.has(path)" is true. Continuing.
   at: _process_graph (scene/animation/animation_tree.cpp:990)
ERROR: Condition "!track_cache.has(path)" is true. Continuing.
   at: _process_graph (scene/animation/animation_tree.cpp:990)
ERROR: Condition "!track_cache.has(path)" is true. Continuing.
   at: _process_graph (scene/animation/animation_tree.cpp:990)
ERROR: Condition "!track_cache.has(path)" is true. Continuing.
   at: _process_graph (scene/animation/animation_tree.cpp:990)
ERROR: Condition "!track_cache.has(path)" is true. Continuing.
   at: _process_graph (scene/animation/animation_tree.cpp:990)
...
```

I ended up closing the application because it would not resolve.

### Steps to reproduce

I was updating the AnimationTree, with the graph open, and changing the referenced tracks from the AnimationPlayer. I ended up saving the scene with the AnimationTree graph open. 

Now, when ever I launch the scene, it throws this infinite loop and makes my machine come to a crawl.

The first output I notice when opening the scene is the following:

```
ERROR: Node not found: "Root/Skeleton3D:root.x" (relative to "/root/@@16122/@@633/@@634/@@643/@@645/@@651/@@657/@@658/@@659/@@681/@@682/@@721/MainScreen/@@9499/@@9328/@@9329/@@9330/@@9331/@@9332/@@9333/Player/Root").
   at: get_node (scene/main/node.cpp:1356)
ERROR: AnimationTree: 'Player/Idle', couldn't resolve track:  'Root/Skeleton3D:root.x'
   at: _update_caches (scene/animation/animation_tree.cpp:587)
ERROR: Node not found: "Root/Skeleton3D:root.x" (relative to "/root/@@16122/@@633/@@634/@@643/@@645/@@651/@@657/@@658/@@659/@@681/@@682/@@721/MainScreen/@@9499/@@9328/@@9329/@@9330/@@9331/@@9332/@@9333/Player/Root").
   at: get_node (scene/main/node.cpp:1356)
ERROR: AnimationTree: 'Player/Idle', couldn't resolve track:  'Root/Skeleton3D:root.x'
   at: _update_caches (scene/animation/animation_tree.cpp:587)
ERROR: Node not found: "Root/Skeleton3D:foot.l" (relative to "/root/@@16122/@@633/@@634/@@643/@@645/@@651/@@657/@@658/@@659/@@681/@@682/@@721/MainScreen/@@9499/@@9328/@@9329/@@9330/@@9331/@@9332/@@9333/Player/Root").
   at: get_node (scene/main/node.cpp:1356)
ERROR: AnimationTree: 'Player/Idle', couldn't resolve track:  'Root/Skeleton3D:foot.l'
   at: _update_caches (scene/animation/animation_tree.cpp:587)
ERROR: Node not found: "Root/Skeleton3D:foot.l" (relative to "/root/@@16122/@@633/@@634/@@643/@@645/@@651/@@657/@@658/@@659/@@681/@@682/@@721/MainScreen/@@9499/@@9328/@@9329/@@9330/@@9331/@@9332/@@9333/Player/Root").
   at: get_node (scene/main/node.cpp:1356)
ERROR: AnimationTree: 'Player/Idle', couldn't resolve track:  'Root/Skeleton3D:foot.l'
   at: _update_caches (scene/animation/animation_tree.cpp:587)
ERROR: Node not found: "Root/Skeleton3D:toes_01.l" (relative to "/root/@@16122/@@633/@@634/@@643/@@645/@@651/@@657/@@658/@@659/@@681/@@682/@@721/MainScreen/@@9499/@@9328/@@9329/@@9330/@@9331/@@9332/@@9333/Player/Root").
   at: get_node (scene/main/node.cpp:1356)
ERROR: AnimationTree: 'Player/Idle', couldn't resolve track:  'Root/Skeleton3D:toes_01.l'
   at: _update_caches (scene/animation/animation_tree.cpp:587)
ERROR: Node not found: "Root/Skeleton3D:thigh_twist.l" (relative to "/root/@@16122/@@633/@@634/@@643/@@645/@@651/@@657/@@658/@@659/@@681/@@682/@@721/MainScreen/@@9499/@@9328/@@9329/@@9330/@@9331/@@9332/@@9333/Player/Root").
   at: get_node (scene/main/node.cpp:1356)
ERROR: AnimationTree: 'Player/Idle', couldn't resolve track:  'Root/Skeleton3D:thigh_twist.l'
   at: _update_caches (scene/animation/animation_tree.cpp:587)
ERROR: Node not found: "Root/Skeleton3D:thigh_twist.l" (relative to "/root/@@16122/@@633/@@634/@@643/@@645/@@651/@@657/@@658/@@659/@@681/@@682/@@721/MainScreen/@@9499/@@9328/@@9329/@@9330/@@9331/@@9332/@@9333/Player/Root").
   at: get_node (scene/main/node.cpp:1356)
ERROR: AnimationTree: 'Player/Idle', couldn't resolve track:  'Root/Skeleton3D:thigh_twist.l'
   at: _update_caches (scene/animation/animation_tree.cpp:587)
ERROR: Node not found: "Root/Skeleton3D:foot.r" (relative to "/root/@@16122/@@633/@@634/@@643/@@645/@@651/@@657/@@658/@@659/@@681/@@682/@@721/MainScreen/@@9499/@@9328/@@9329/@@9330/@@9331/@@9332/@@9333/Player/Root").
   at: get_node (scene/main/node.cpp:1356)
ERROR: AnimationTree: 'Player/Idle', couldn't resolve track:  'Root/Skeleton3D:foot.r'
   at: _update_caches (scene/animation/animation_tree.cpp:587)
ERROR: Node not found: "Root/Skeleton3D:foot.r" (relative to "/root/@@16122/@@633/@@634/@@643/@@645/@@651/@@657/@@658/@@659/@@681/@@682/@@721/MainScreen/@@9499/@@9328/@@9329/@@9330/@@9331/@@9332/@@9333/Player/Root").
   at: get_node (scene/main/node.cpp:1356)
...
```

The above output, goes through all the bones on the skeleton...

Which then goes into the mentioned error output of:

```
ERROR: Condition "!track_cache.has(path)" is true. Continuing.
   at: _process_graph (scene/animation/animation_tree.cpp:990)
ERROR: Condition "!track_cache.has(path)" is true. Continuing.
   at: _process_graph (scene/animation/animation_tree.cpp:990)
ERROR: Condition "!track_cache.has(path)" is true. Continuing.
   at: _process_graph (scene/animation/animation_tree.cpp:990)
ERROR: Condition "!track_cache.has(path)" is true. Continuing.
   at: _process_graph (scene/animation/animation_tree.cpp:990)
ERROR: Condition "!track_cache.has(path)" is true. Continuing.
   at: _process_graph (scene/animation/animation_tree.cpp:990)
ERROR: Condition "!track_cache.has(path)" is true. Continuing.
   at: _process_graph (scene/animation/animation_tree.cpp:990)
ERROR: Condition "!track_cache.has(path)" is true. Continuing.
   at: _process_graph (scene/animation/animation_tree.cpp:990)
ERROR: Condition "!track_cache.has(path)" is true. Continuing.
   at: _process_graph (scene/animation/animation_tree.cpp:990)
ERROR: Condition "!track_cache.has(path)" is true. Continuing.
   at: _process_graph (scene/animation/animation_tree.cpp:990)
ERROR: Condition "!track_cache.has(path)" is true. Continuing.
   at: _process_graph (scene/animation/animation_tree.cpp:990)
ERROR: Condition "!track_cache.has(path)" is true. Continuing.
   at: _process_graph (scene/animation/animation_tree.cpp:990)
ERROR: Condition "!track_cache.has(path)" is true. Continuing.
   at: _process_graph (scene/animation/animation_tree.cpp:990)
ERROR: Condition "!track_cache.has(path)" is true. Continuing.
   at: _process_graph (scene/animation/animation_tree.cpp:990)
ERROR: Condition "!track_cache.has(path)" is true. Continuing.
   at: _process_graph (scene/animation/animation_tree.cpp:990)
ERROR: Condition "!track_cache.has(path)" is true. Continuing.
   at: _process_graph (scene/animation/animation_tree.cpp:990)
ERROR: Condition "!track_cache.has(path)" is true. Continuing.
   at: _process_graph (scene/animation/animation_tree.cpp:990)
ERROR: Condition "!track_cache.has(path)" is true. Continuing.
   at: _process_graph (scene/animation/animation_tree.cpp:990)
ERROR: Condition "!track_cache.has(path)" is true. Continuing.
   at: _process_graph (scene/animation/animation_tree.cpp:990)
ERROR: Condition "!track_cache.has(path)" is true. Continuing.
   at: _process_graph (scene/animation/animation_tree.cpp:990)
ERROR: Condition "!track_cache.has(path)" is true. Continuing.
   at: _process_graph (scene/animation/animation_tree.cpp:990)
ERROR: Condition "!track_cache.has(path)" is true. Continuing.
   at: _process_graph (scene/animation/animation_tree.cpp:990)
ERROR: Condition "!track_cache.has(path)" is true. Continuing.
   at: _process_graph (scene/animation/animation_tree.cpp:990)
...
```

This happens until I close the application, which takes a long time because my machine is locked up.

### Minimal reproduction project

_No response_

### PR DESCRIPTION
**Pull request by TokageItLab**
Fixes #65608
Supersedes #68998

There are two reasons for the #65608 spamming:

First, when caching tracks, if they contain invalid paths, they throw an error and are not cached. The iterating is done on the list of tracks, not on the list of caches. If the cache is not found when iterating on a track, AnimationTree throws that error every frame.

#68998 tried to remove that error, but I didn't really think that was a good way. However, after I looked into AnimationPlayer, it doesn't throw errors in the same case, so it seems to be acceptable, I apologize for that.

However, there is another cause for spamming. If a large number of animations and a large number of track paths cannot be resolved, a large number of errors can occur during caching. This means that this can happen frequently with 3D skeletal animations.

Considering the existence of retargeting and duck typing recommendations, it might be fine that not to throw errors even for invalid paths. So I added an option to Project Setting to disable checking for paths in AnimationPlayer/AnimationTree.

image

### CODE REVIEW

**Comment by fire** [comment]
>  disable checking for paths in AnimationPlayer/AnimationTree.

I think being able to playback animations without halting is better than catching errors.

It feels less of a landmine.

File: doc/classes/ProjectSettings.xml [hunk 1]
```diff
@@ -188,6 +188,12 @@
 		</method>
 	</methods>
 	<members>
+		<member name="animation/animation_player/warnings/ignore_invalid_paths" type="bool" setter="" getter="" default="true">
+			If [code]true[/code], [AnimationPlayer] doesn't throw the error of that there is no object of the track path in the scene.
```
Comment (fire) [inline doc/classes/ProjectSettings.xml:192]:
```suggestion
			If [code]true[/code], [AnimationPlayer] doesn't throw the error that there is no object of the track path in the scene.
```

File: doc/classes/ProjectSettings.xml [hunk 2]
```diff
@@ -188,6 +188,12 @@
...
+		</member>
+		<member name="animation/animation_tree/warnings/ignore_invalid_paths" type="bool" setter="" getter="" default="true">
+			If [code]true[/code], [AnimationTree] doesn't throw the error of that there is no object of the track path in the scene.
```
Comment (fire) [inline doc/classes/ProjectSettings.xml:195]:
```suggestion
			If [code]true[/code], [AnimationTree] doesn't throw the error that there is no object of the track path in the scene.
```

**Review by fire — APPROVED** [review]
Some minor nitpicks that should be resolved, but we discussed out of the thread.

File: doc/classes/ProjectSettings.xml [hunk 3]
```diff
@@ -188,6 +188,12 @@
 		</method>
 	</methods>
 	<members>
+		<member name="animation/animation_player/warnings/ignore_invalid_paths" type="bool" setter="" getter="" default="true">
+			If [code]true[/code], [AnimationPlayer] doesn't throw the error that there is no object of the track path in the scene.
```
Comment (fire) [inline doc/classes/ProjectSettings.xml:192]:
```suggestion
			If [code]true[/code], [AnimationPlayer] silences the error of no matching object of the track path in the scene.
```

**Comment by TokageItLab** [comment]
There is a problem with the path checking of subpaths, I'm fixing...

**Comment by TokageItLab** [comment]
Okay fixed. Checking path needs to use `has_node()`, not `has_node_and_resource()`.

**Comment by akien-mga** [comment]
I'm not convinced that this is the right approach.

First of all the added project settings are a bit awkward (new "Animation" top level category just for two new sections with just one equivalent setting each... that doesn't look good user friendly).

And making logic dependent on a setting most user won't even find also doesn't seem satisfactory. Users will still get spammed, report a bug, be told it's a feature and they can disable it, and then possibly miss situations where things are actually warning for a reason they need to be aware of.

**Comment by TokageItLab** [comment]
Well yes, Honestly, I am not so much in favor of hiding these errors. We should be a bit more cautious.

**Comment by fire** [comment]
I prefer no warning over halting the game for a few frames, because you can't debug halting.

**Comment by mhilbrunner** [comment]
Discussed in animation, we kind of agreed the project setting isn't really a great solution. We need to rethink this one.

**Comment by akien-mga** [comment]
Superseded by #86608.

**Pull request closed without merging**